SabreTools.Serialization/Readers/ISO9660.cs

```diff
@@ -0,0 +1,799 @@
+        public static BootRecordVolumeDescriptor? ParseBootRecordVolumeDescriptor(Stream data, short sectorLength)
+        {
+            var obj = new BootRecordVolumeDescriptor();
+
+            obj.Type = VolumeDescriptorType.VOLUME_PARTITION_DESCRIPTOR;
+            obj.Identifier = data.ReadBytes(5);
+
+            // Validate Identifier, return null and rewind if invalid
+            if (!obj.Identifier.EqualsExactly(Constants.StandardIdentifier))
+            {
+                data.Position -= 6;
+                return null;
+            }
+
+            obj.Version = data.ReadByteValue();
+            obj.BootSystemIdentifier = data.ReadBytes(32);
+            obj.BootIdentifier = data.ReadBytes(32);
+            obj.BootSystemUse = data.ReadBytes(1977);
+
+            // Skip remainder of the logical sector
+            if (sectorLength > Constants.MinimumSectorSize)
+                data.Position += sectorLength - Constants.MinimumSectorSize;
+
+            return obj;
+        }
+
+        /// <summary>
+        /// Parse a Stream into a PrimaryVolumeDescriptor
+        /// </summary>
+        /// <param name="data">Stream to parse</param>
+        /// <param name="sectorLength">Number of bytes in a logical sector (usually 2048)</param>
+        /// <returns>Filled PrimaryVolumeDescriptor on success, null on error</returns>
+        public static PrimaryVolumeDescriptor? ParsePrimaryVolumeDescriptor(Stream data, short sectorLength)
+        {
+            var obj = new PrimaryVolumeDescriptor();
+
+            obj.Type = VolumeDescriptorType.PRIMARY_VOLUME_DESCRIPTOR;
+            return (PrimaryVolumeDescriptor?)ParseBaseVolumeDescriptor(data, sectorLength, obj);
+        }
+
+        /// <summary>
+        /// Parse a Stream into a SupplementaryVolumeDescriptor
+        /// </summary>
+        /// <param name="data">Stream to parse</param>
+        /// <param name="sectorLength">Number of bytes in a logical sector (usually 2048)</param>
+        /// <returns>Filled SupplementaryVolumeDescriptor on success, null on error</returns>
+        public static SupplementaryVolumeDescriptor? ParseSupplementaryVolumeDescriptor(Stream data, short sectorLength)
+        {
+            var obj = new SupplementaryVolumeDescriptor();
+
+            obj.Type = VolumeDescriptorType.SUPPLEMENTARY_VOLUME_DESCRIPTOR;
+            return (SupplementaryVolumeDescriptor?)ParseBaseVolumeDescriptor(data, sectorLength, obj);
+        }
+
+        /// <summary>
+        /// Parse a Stream into a BaseVolumeDescriptor
+        /// </summary>
+        /// <param name="data">Stream to parse</param>
+        /// <param name="sectorLength">Number of bytes in a logical sector (usually 2048)</param>
+        /// <returns>Filled BaseVolumeDescriptor on success, null on error</returns>
+        public static BaseVolumeDescriptor? ParseBaseVolumeDescriptor(Stream data, short sectorLength, BaseVolumeDescriptor obj)

```

Comment:
I'll admit that this feels like a bit of an odd way of handling the base type situation. I don't really have a good alternative to this approach at this time, but I will try to think on it. If I can't figure out any alternative, then it can stay this way.